cmd/bccatomicswap/main.go

```diff
@@ -0,0 +1,1133 @@
+	}
+
+	connConfig := &rpc.ConnConfig{
+		Host:         connect,
+		User:         *rpcuserFlag,
+		Pass:         *rpcpassFlag,
+		DisableTLS:   true,
+		HTTPPostMode: true,
+	}
+	client, err := rpc.New(connConfig, nil)
+	if err != nil {
+		return fmt.Errorf("rpc connect: %v", err), false
+	}
+	defer func() {
+		client.Shutdown()
+		client.WaitForShutdown()
+	}()
+
+	err = cmd.runCommand(client)
+	return err, false
+}
+
+func normalizeAddress(addr string, defaultPort string) (hostport string, err error) {
+	host, port, origErr := net.SplitHostPort(addr)
+	if origErr == nil {
+		return net.JoinHostPort(host, port), nil
+	}
+	addr = net.JoinHostPort(addr, defaultPort)
+	_, _, err = net.SplitHostPort(addr)
+	if err != nil {
+		return "", origErr
+	}
+	return addr, nil
+}
+
+func walletPort(params *chaincfg.Params) string {
+	switch params {
+	case &chaincfg.MainNetParams:
+		return "8332"
+	case &chaincfg.TestNet3Params:
+		return "18332"
+	default:
+		return ""
+	}
+}
+
+// createSig creates and returns the serialized raw signature and compressed
+// pubkey for a transaction input signature.  Due to limitations of the Bitcoin
+// Core RPC API, this requires dumping a private key and signing in the client,
+// rather than letting the wallet sign.
+func createSig(tx *wire.MsgTx, idx int, pkScript []byte, addr btcutil.Address,
+	c *rpc.Client, val int64) (sig, pubkey []byte, err error) {
+	wif, err := c.DumpPrivKey(addr)
+	if err != nil {
+		return nil, nil, err
+	}
+	sig, err = bchutil.RawTxInSignature(tx, idx, pkScript, txscript.SigHashAll, wif.PrivKey, val)
+	if err != nil {
+		return nil, nil, err
+	}
+	return sig, wif.PrivKey.PubKey().SerializeCompressed(), nil
+}
+
+// fundRawTransaction calls the fundrawtransaction JSON-RPC method.  It is
+// implemented manually as client support is currently missing from the
+// btcd/rpcclient package.
+func fundRawTransaction(c *rpc.Client, tx *wire.MsgTx, feePerKb btcutil.Amount) (fundedTx *wire.MsgTx, fee btcutil.Amount, err error) {
+	var buf bytes.Buffer
+	buf.Grow(tx.SerializeSize())
+	tx.Serialize(&buf)
+	param0, err := json.Marshal(hex.EncodeToString(buf.Bytes()))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	param1, err := json.Marshal(struct {
+		FeeRate float64 `json:"feeRate"`
+	}{
+		FeeRate: feePerKb.ToBTC(),
+	})
+	if err != nil {
+		return nil, 0, err
+	}
+
+	params := []json.RawMessage{param0, param1}
+	rawResp, err := c.RawRequest("fundrawtransaction", params)
+	if err != nil {
+		rawResp, err := c.RawRequest("fundrawtransaction", []json.RawMessage{param0})
+		if err != nil {
+			return nil, 0, err
+		}
+		fmt.Printf("warning: your node (probably Bitcoin Classic or Bitcoin Unlimited) doesn't support setting fees manually. Consider using Bitcoin ABC or XT.")
+		return fundTransactionViaRPC(rawResp)
+	}
+	return fundTransactionViaRPC(rawResp)
+}
+
+func fundTransactionViaRPC(rawResp json.RawMessage) (fundedTx *wire.MsgTx, fee btcutil.Amount, err error) {
```

Comment:
this function name is sort of weird, since the other func above this one is the one actually doing the RPC, and this is just working with a json object.